File: linked-list/python/ReverseLinkedList.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

def reverseList(head):
    prev, curr = None, head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    head = prev
    return head

if __name__=='__main__':
    head = Node(10)
    head.next = Node(20)
    head.next.next = Node(30)
    head.next.next.next = Node(40)
    head.next.next.next.next = Node(50)
    head = reverseList(head)
    while head:
        print(head.val, end = ' -> ')
        head = head.next
    print('X')

File: linked-list/java/practice2/ReverseLinkedList.java
package practice2;

public class ReverseLinkedList {

    static class Node {
        int data;
        Node next;

        public Node(int data) {
            this.data = data;
        }
    }

    public static void display(Node node) {
        while (node != null) {
            System.out.print(node.data+ " ");
            node = node.next;
        }
        System.out.println("");
    }

    public Node reverseList(Node head) {
        Node prev = null;
        Node post = null;
        Node ptr = head;
        while (ptr != null) {
            post = ptr.next;
            ptr.next = prev;
            prev = ptr;
            ptr = post;
        }
        return prev;
    }

    public static void main(String[] args) {
        Node head = new Node(1);
        head.next = new Node(2);
        head.next.next = new Node(3);
        head.next.next.next = new Node(4);
        head.next.next.next.next = new Node(5);
        head.next.next.next.next.next = new Node(6);
        head.next.next.next.next.next.next = new Node(7);

        display(head);
        ReverseLinkedList solution = new ReverseLinkedList();
        Node reversedHead = solution.reverseList(head);
        display(reversedHead);

    }
}
